Feature/dynamic import preloading (#361)
GoogleChromeLabs/tooling.report: tests/code-splitting/subtests/dynamic-import-preloading/index.md

---
title: Dynamic import Preloading
importance: 1
shortDesc: 'Can code splitting be preloaded?'
---

# Introduction

While loading the module which has been code spllited, it is important to load its dependencies in parallel to
optimize the load time.

# The Test

This test checks to see if it's possible to start network request for the dependecies as soon as the code splitted module start loading.

> **index.js**
> 
> ```js
> setTimeout(async () => {
>   const { shout } = await import('./utils.js');
>   shout('this is index');
> }, 6000);
> ```
> 
> **utils.js**
> 
> ```js
> import { exclaim } from './exclaim.js';
> 
> export function logCaps(msg) {
>   if (!msg) return;
>   console.log(msg.toUpperCase());
> }
> 
> export function shout(msg) {
>   logCaps(exclaim(msg));
> }
> ```

Reviewer: It's easier to trigger this behavior with two on-demand loaded modules that reference a common module:

`````suggestion
**index.js**

```js
setTimeout(async () => {
  const { shout } = Math.random() < 0.5 ? await import('./utilsA.js') : await import('./utilsB.js');
  shout('this is index');
}, 6000);
```

**utilsA.js**

```js
import { exclaim } from './exclaim.js';

export function logCaps(msg) {
  if (!msg) return;
  console.log(msg.toUpperCase());
}

export function shout(msg) {
  logCaps(exclaim(msg));
}
```

**utilsB.js**

```js
import { exclaim } from './exclaim.js';

export function logNormal(msg) {
  if (!msg) return;
  console.log(msg);
}

export function shout(msg) {
  logNormal(exclaim(msg));
}
```
`````

---

Rendered:

**index.js**

```js
setTimeout(async () => {
  const { shout } = Math.random() < 0.5 ? await import('./utilsA.js') : await import('./utilsB.js');
  shout('this is index');
}, 6000);
```

**utilsA.js**

```js
import { exclaim } from './exclaim.js';

export function logCaps(msg) {
  if (!msg) return;
  console.log(msg.toUpperCase());
}

export function shout(msg) {
  logCaps(exclaim(msg));
}
```

**utilsB.js**

```js
import { exclaim } from './exclaim.js';

export function logNormal(msg) {
  if (!msg) return;
  console.log(msg);
}

export function shout(msg) {
  logNormal(exclaim(msg));
}
```

Author: Thanks for the suggestion.

---

> title: Dynamic import Preloading
> importance: 1
> shortDesc: 'Can code splitting be preloaded?'

Reviewer 2: ```suggestion
title: Dynamic import preloading
importance: 1
shortDesc: 'Can lazy-loaded split points be preloaded?'
```

Reviewer: I find both titles confusing. This sounds like related to `meta preload`. But instead this test is related to on demand loading efficiency.

Author: I agree with @sokra . This test's title is misleading, it implies that there must be some sort of preloading is involved via preload meta tag. I think something like `On-demand loading of code splitted modules` would make it more clearer.

---

# Introduction

> While loading the module which has been code spllited, it is important to load its dependencies in parallel to
> optimize the load time.

Reviewer 2: ```suggestion
When creating split points [using dynamic import](/code-splitting/dynamic-import/), it is possible to construct a "waterfall" scenario where JavaScript bundles depend on additional bundles discovered once executed. The ability to preload nested dependency bundles is important in these cases, since doing so can allow the dependency bundles to be downloaded and evaluated in parallel.
```

# The Test

> This test checks to see if it's possible to start network request for the dependecies as soon as the code splitted module start loading.
> **index.js**

Reviewer 2: ```suggestion
This test checks to see if it's possible to request code-splitted dependency bundles before the consuming bundle has been loaded, either via [preload](https://web.dev/preload-critical-assets/) or async script loading.

**index.js**
```

Reviewer: I think you both talk about different tests that both make sense.

Preloading/Prefetching: Load files related to a `import()` before they are really requested.

Parallel loading of dependencies: Does an `import()` cause multiple parallel requests or an waterfall of requests?

```js
setTimeout(async () => {
  const { shout } =
    Math.random() < 0.5
      ? await import('./utilsA.js')

>       : await import('./utilsB.js');

Reviewer 2: Seems like this could just be a conditional import of one `utils.js` module with the dice roll - it just needs to not be an unconditional import.

Author: @developit I don't get it.

Reviewer: Lgtm but I would go with `random() < 0` so that the execution is deterministic. Just `false` doesn't work as this is usually detected by bundlers.

>     Math.random() < 0.5

Reviewer 3: I can’t find an explanation in this README for this `random()`? What is this for?

Reviewer: I proposed that, but you can also use other conditions. In the end you need to `import()` two different modules that share a common dependency. This way you create a chunking where one output depends on another one that can be loaded in parallel or preloaded.

Reviewer 3: OOoh, it’s just so we have 2 modules with a shared dependency. Got it. It warrants an explanation or even simpler: just 2 consecutive `import()` calls.

Reviewer: That would also work. I proposed that to simulate an A/B test scenario.

Reviewer 3: That makes sense. A short comment to explain this would be good, tho. (cc @AjayPoshak)

---
title: Dynamic import preloading
importance: 1
shortDesc: 'Can lazy-loaded split points be preloaded?'
---

# Introduction

When creating split points [using dynamic import](/code-splitting/dynamic-import/), it is possible to construct a "waterfall" scenario where JavaScript bundles depend on additional bundles discovered once executed. The ability to preload nested dependency bundles is important in these cases, since doing so can allow the dependency bundles to be downloaded and evaluated in parallel.

# The Test

This test checks to see if it's possible to request code-splitted dependency bundles before the consuming bundle has been loaded, either via [preload](https://web.dev/preload-critical-assets/) or async script loading.

**index.js**

```js
(async () => {
  const { shout } =

>     Math.random() < 0.5
>       ? await import('./utilsA.js')
>       : await import('./utilsB.js');

Reviewer 3: ```suggestion
	// `utilsA` and `utilsB` share a dependency.
	// One import is conditional to make sure the imports and the
	// common dependency stay separate files and will cause a 
	// waterfall loading pattern when loaded naïvely.
    import('./utilsA.js')
    if (Math.random() < 0.5) {
      import('./utilsB.js');
    }
```

@AjayPoshak @developit What do you think about this as a variant/explanation to address the comment above?

Reviewer 3: (if this is okay with both of you, can you  adjust the samples accordingly?)

Reviewer 2: Originally I was going to comment that this seems to make the loading order deterministic, but actually I think that makes the test easier to verify. The correct outcome is that A is always executed prior to B, and in the test this can be verified by execution of B never happening before A.

Otherwise I was going to suggest having the random check swap the order of both imports.

Reviewer 3: I don’t think execution order is well-defined here, since I am not `await`’ing.

Reviewer 3: I guess what I am saying: Both execution orders are valid, which I like because we only want to test if loading happens in parallel. Or am I misunderstanding you?

Author: @surma Should I change make changes according to this snippet?

Reviewer 3: Yes I think so, but I am not quite sure if @developit agrees :D

Author: Lets merge this PR for now, we can move this discussion to issues, and once we have consensus, we can open a PR to address it.  WDYT?? @surma

Reviewer 3: I’d definitely want an explanatory comment in the code sample before switching this live. Happy to defer the discussion _which_ example is better to a separate issue, tho.

Reviewer 2: I'm happy with the commented version as suggested here. TBH I'm not 100% sure what I was trying to say in my earlier comment. I think it may have been related to order preservation when using hoisted `import()` statements rather than preloading them via `<link>`, since that changes execution order whereas preload does not.

Reviewer 3: Cheers Jason! 

@AjayPoshak Alright! Home stretch! Do you mind taking the suggestion above and adjusting the samples? Then we can merge this 🎉

Author: sure, I am on it.

Author: @surma I have pushed my changes. Please review.

  shout('this is index');

> }, 6000);

Reviewer 4: Is the `setTimeout` a necessary part of the test?

Reviewer 3: I don’t believe it is.
